wdl-lint/src/rules/disallowed_declaration_name.rs

```diff
@@ -0,0 +1,375 @@
+        type_names.insert(primitive_type.to_string());
+
+        // For Boolean type, also check for "Bool"
+        if primitive_type.kind() == PrimitiveTypeKind::Boolean {
+            type_names.insert("Bool".to_string());
+        }
+    }
+
+    // Add compound type names
+    if let Some(array_type) = ty.as_array_type() {
+        // Add "Array" for the compound type
+        type_names.insert("Array".to_string());
+
+        // Extract inner type names but handle plurals differently
+        if let Some(primitive) = array_type.element_type().as_primitive_type() {
+            // Skip File and String types
+            match primitive.kind() {
+                PrimitiveTypeKind::File | PrimitiveTypeKind::String => {}
+                _ => {
+                    // Don't check for inner primitive type suffixes in arrays
+                    // since the plural form is often appropriate
+                    // (e.g., Array[Int] integers is fine)
+                }
+            }
+        }
+    } else if let Some(_map_type) = ty.as_map_type() {
+        type_names.insert("Map".to_string());
+
+        // Skip checking inner types of maps
+    } else if let Some(_pair_type) = ty.as_pair_type() {
+        type_names.insert("Pair".to_string());
+
+        // Skip checking inner types of pairs
+    }
+
+    // Check if the declaration name ends with one of the type names
+    for type_name in &type_names {
+        let type_lower = type_name.to_lowercase();
+
+        // Skip if the type name is too short
+        if type_lower.len() < 3 {
+            continue;
+        }
+
+        // Special handling for Int
+        if type_lower == "int" {
+            // Split the identifier into words to check if "int" is used as a standalone suffix
+            let words = split_to_words(name_str);
+
+            // Check if "int" appears as the last word
+            if let Some(last_word) = words.last() {
+                if last_word.to_lowercase() == "int" {
+                    state.exceptable_add(
+                        decl_identifier_with_type(decl.name().span(), name_str, type_name),
+                        SyntaxElement::from(decl.syntax().clone()),
+                        exceptable_nodes,
+                    );
+                    return;
+                }
+            }
+        } else {
+            // For other types, check if the identifier ends with the type name
+            let name_lower = name_str.to_lowercase();
+            if name_lower.ends_with(&type_lower) {
+                state.exceptable_add(
+                    decl_identifier_with_type(decl.name().span(), name_str, type_name),
+                    SyntaxElement::from(decl.syntax().clone()),
+                    exceptable_nodes,
+                );
+                return;
+            }
+        }
+    }
+}
+
+/// Split an identifier into words using convert_case
+fn split_to_words(identifier: &str) -> Vec<String> {
```

Comment:
I think this can be achieved with some builtin `convert_case` functionality, instead of rewriting your self. Check out these two sections from the docs, which I think can be useful here
https://docs.rs/convert_case/latest/convert_case/fn.split.html
https://docs.rs/convert_case/latest/convert_case/struct.Boundary.html#method.defaults